mm2src/mm2_main/src/lp_swap/taker_swap.rs

```diff
@@ -1905,86 +1906,91 @@ impl TakerSwap {
-                    ),
-                ])),
-            };
-        }
-
+        // Keep trying to recover funds (by refunding the taker payment or spending the maker payment) until successful or face an irrecoverable error.
         loop {
-            match self.taker_coin.can_refund_htlc(locktime).await {
-                Ok(CanRefundHtlc::CanRefundNow) => break,
-                Ok(CanRefundHtlc::HaveToWait(to_sleep)) => Timer::sleep(to_sleep as f64).await,
-                Err(e) => {
-                    error!("Error {} on can_refund_htlc, retrying in 30 seconds", e);
-                    Timer::sleep(30.).await;
+            match self.recover_funds().await {
+                // We recovered the swap successfully.
+                Ok(recovered_swap) => match recovered_swap {
+                    // We recovered the swap by refunding the taker payment.
+                    RecoveredSwap {
+                        action: RecoveredSwapAction::RefundedMyPayment,
+                        coin: _,
+                        transaction,
+                    } => {
+                        let tx_ident = TransactionIdentifier {
+                            tx_hex: transaction.tx_hex().into(),
+                            tx_hash: transaction.tx_hash_as_bytes(),
+                        };
+                        info!("Taker payment refund tx {:02x}", tx_ident.tx_hash);
+                        return Ok((Some(TakerSwapCommand::FinalizeTakerPaymentRefund), vec![
+                            TakerSwapEvent::TakerPaymentRefunded(Some(tx_ident)),
+                        ]));
+                    },
+                    // We recovered the swap by proceeding forward and spending the maker payment. The swap wasn't actually a failure.
+                    // Roll back to confirming the maker payment spend.
+                    RecoveredSwap {
+                        action: RecoveredSwapAction::SpentOtherPayment,
+                        coin: _,
+                        transaction,
+                    } => {
+                        let tx_ident = TransactionIdentifier {
+                            tx_hex: transaction.tx_hex().into(),
+                            tx_hash: transaction.tx_hash_as_bytes(),
+                        };
+                        info!("Refund canceled. Taker payment spend tx {:02x}", tx_ident.tx_hash);
+                        // TODO: We prepared for refund but didn't finalize refund. This must be breaking something for lightning.
+                        // We better find a way to rollback the state machine and remove erroneous events,
+                        // the swap at this point will be marked as errored but in fact it recovered from the error.
+                        return Ok((Some(TakerSwapCommand::ConfirmMakerPaymentSpend), vec![
+                            TakerSwapEvent::MakerPaymentSpent(tx_ident),
+                        ]));
+                    },
```

Comment:
Also can be simplified with closure and `return match`

```rust
                // We recovered the swap successfully.
                Ok(recovered_swap) => {
                    let tx_ident = |transaction: TransactionEnum| TransactionIdentifier {
                        tx_hex: transaction.tx_hex().into(),
                        tx_hash: transaction.tx_hash_as_bytes(),
                    };

                    return match recovered_swap.action {
                        // We recovered the swap by refunding the taker payment.
                        RecoveredSwapAction::RefundedMyPayment => {
                            let ident = tx_ident(recovered_swap.transaction);
                            info!("Taker payment refund tx {:02x}", ident.tx_hash);
                            Ok((Some(TakerSwapCommand::FinalizeTakerPaymentRefund), vec![
                                TakerSwapEvent::TakerPaymentRefunded(Some(ident)),
                            ]))
                        },
                        // We recovered the swap by proceeding forward and spending the maker payment. The swap wasn't actually a failure.
                        // Roll back to confirming the maker payment spend.
                        RecoveredSwapAction::SpentOtherPayment => {
                            let ident = tx_ident(recovered_swap.transaction);
                            info!("Refund canceled. Taker payment spend tx {:02x}", ident.tx_hash);
                            // TODO: We prepared for refund but didn't finalize refund. This must be breaking something for lightning.
                            // We better find a way to rollback the state machine and remove erroneous events,
                            // the swap at this point will be marked as errored but in fact it recovered from the error.
                            Ok((Some(TakerSwapCommand::ConfirmMakerPaymentSpend), vec![
                                TakerSwapEvent::MakerPaymentSpent(ident),
                            ]))
                        },
                    };
                },
```

Reply:
aha i see,
you could also run tx_ident outside the match since it's the exact same code for both branches.
i did it this way though since it fits more with the style of the error routes so i thought it will be easier to read. sometimes i lean for readability & style and ignore other aspects :)
but sense u mentioned it, i guess i looks off, so let's combine the combinable code outside the match then.

Reply:
[5e0709d](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2280/commits/5e0709d22a683c3304019872885aeaf438feb20f)